osu.Game/Modes/ScoreProcesssor.cs

```diff
@@ -46,6 +62,11 @@ public void AddJudgement(JudgementInfo judgement)
             UpdateCalculations(judgement);
 
             judgement.ComboAtHit = (ulong)Combo.Value;
+            if (Health.Value == Health.MinValue && !HasFailed)
+            {
+                HasFailed = true;
+                Failed?.Invoke();

```

Comment:
i've made HasFailed's set private instead. I don't think it should be changed externally for now.